README.md

# Indiana COVID-19 Tracker

![Noon
Update](https://github.com/ercbk/Indiana-COVID-19-Tracker/workflows/Noon%20Update/badge.svg)
![Evening
Update](https://github.com/ercbk/Indiana-COVID-19-Tracker/workflows/Evening%20Update/badge.svg)

Charts updated between noon and 1pm ET and at 6:30 pm ET  
Changelog located in
[NEWS.md](https://github.com/ercbk/Indiana-COVID-19-Tracker/blob/master/NEWS.md)

This repository tracks COVID-19 data statewide and by county. Most of
the State and County data for positive tests and deaths are gathered
from The New York Times COVID-19
[repository](https://github.com/nytimes/covid-19-data) and The Indiana
Data [Hub](https://hub.mph.in.gov/dataset).

Governor Holcomb’s reopening plan began on May 4th, 2020. The plan
resembled the White House
[plan](https://www.whitehouse.gov/openingamerica/) in many aspects, and
both were simliar to the
[guidelines](https://www.aei.org/research-products/report/national-coronavirus-response-a-road-map-to-reopening/)
published by the American Enterpise Institute (AEI) and Johns Hopkins.
But neither Governor Holcomb’s nor the White House’s plan included the
snapback (i.e. reverting to a previous stage) triggering conditions
listed in the AEI proposal. Depending on the data that’s available, I’ll
also be trying to track some of the triggers in the AEI proposal and
those from Governor Holcomb if any are presented.  
<br /> <br /> <br />

<img src="plots/ind-combo-line-2020-12-31.png" width="935" />

  - Each point is the total cases that have accumulated until that date.
    The lines show the exponential nature of the spread of the virus at
    the beginning of the epidemic and changes in trend as policy and
    behavior changes.  
  - *Doubling time* is the amount of time it would take for the current
    total to double. It assumes that the rate of change is constant. The
    doubling time shown here assumes that the total change from
    yesterday to today (i.e. single day *growth rate*) will remain
    constant in the future. It’s not a valid assumption in this
    situation, but it’s interesting as a hypothetical.  
    <br /> <br /> <br />

<img src="plots/pos-policy-line-2020-12-31.png" width="935" />

  - This chart shows the number of daily positive cases on the y-axis
    versus the cumulative total of positive cases on the x-axis.  
  - In the AEI reopening guidelines, one of the triggers for returning
    to stage one is five or more consecutive days where the value of
    daily cases has increased.  
  - Details on the various stages of re-opening can be found at
    [Back-on-Track Indiana](https://backontrack.in.gov/2348.htm).  
    <br /> <br /> <br />

<img src="plots/hosp-icu-vent-2020-12-31.png" width="935" />

  - Even though Governor Holcomb hasn’t explicitly stated the conditions
    that would necessitate a reversal of course to a previous stage, he
    did mention some benchmarks when discussing his guiding principles
    in his re-opening [speech](https://backontrack.in.gov/2362.htm). By
    negating a few of these, we can infer what some of the snapback
    conditions might entail.
      - An increase for 14 consecutive days in the daily total of those
        hospitalized for COVID-19.  
      - The proportion of available ICU beds to total ICU beds being
        below 40% for 14 consecutive days.
      - The proportion of available ventilators to total ventilators
        being below 70% for 14 consecutive days.  
        <br /> <br /> <br />

<img src="plots/pos-rate-line-2020-12-29.png" width="935" />

  - The Positive Test Rate is the number of positive test results
    divided by the number of tests administered over a period of time.
  - During a [briefing](https://twitter.com/i/broadcasts/1ypKdwaqLegxW)
    with Gov. Holcomb and Dr. Box, it was pretty clear that they pay
    close attention to positivity rates and use a seven day window for
    their calculation along with a target rate of less than 5%.
  - Indiana Data Hub tends to revise its counts as it continues to
    collect more data. The rate is calculated over a seven day window,
    so it should be pretty consistent. Even so, it would be prudent not
    to assess the last couple rates too confidently as they are likely
    to change.  
  - U.S. average calculated using data from [The COVID Tracking
    Project](https://covidtracking.com/data)  
    <br /> <br /> <br />

<img src="plots/ot-rest-line-NA.png" width="935" />

  - OpenTable is a service for reviewing and making reservations at
    restaurants. On its “The state of the restaurant industry”
    [page](https://www.opentable.com/state-of-industry), it provides a
    couple different datasets that use reservation and food order data
    to summarize the state of the restaurant industry regionally by
    city, state, and country.  
  - The dataset being visualized uses reservation data from a sample of
    restaurants across Indiana. Each data point is the median daily
    percent difference in seated diners from the previous year to this
    year. So, if the day is the first Tuesday in June 2020, then the
    comparison is between that day and the first Tuesday of June 2019.  
  - At stages two and three, restaurants are allowed to reopen at 50%
    capacity. So, if a restaurant was fully-booked on a specific day in
    2019 and fully booked on that same day in 2020, then we’d still see
    a 50% difference in seated diners for that one restaurant during
    stages two and three. In stage four, restaurants can open to 75%
    capacity, so in the same scenario above, the fully-booked, percent
    difference benchmark becomes 25%. Numbers around these benchmarks
    could indicate the level at which hoosiers are responding after each
    reopening stage.
  - Weekends are shown as vertical bars, and for comparison, the
    horizontal dashed line represents Indiana’s current percent
    difference.  
    <br /> <br /> <br />

<img src="plots/goog-mob-line-2020-12-25.png" width="935" />

  - Google like Apple has released some of the
    [data](https://www.google.com/covid19/mobility/) it collects from
    its Maps app that it uses to produce Community Mobility Reports and
    product recommendations.  
  - From the “About this Data” section in the Report:
      - These reports show how visits and length of stay at different
        places change compared to a baseline.
      - The baseline is the median value, for the corresponding day of
        the week, during the 5- week period Jan 3–Feb 6, 2020.  
  - Weekends are shown as vertical bars.  
    <br /> <br /> <br />

<img src="plots/region-dea-line-2020-12-31.png" width="935" />

<img src="plots/region-pos-line-2020-12-31.png" width="935" />

  - 7-day averages *greater* than 14 day averages indicate an
    *increasing* trend of positive cases or deaths.  
  - 7-day averages *less* than 14 day averages indicate a *decreasing*
    trend in positive cases or deaths.  
  - State policy
    [data](https://docs.google.com/spreadsheets/d/1zu9qEWI8PsOI_i8nI_S29HDGHlIp2lfVMsGxpQ5tvAQ/edit#gid=0)
    comes from Julia Raifman, Kristen Nocka, et al at Boston
    University.  
    <br /> <br /> <br />

<img src="plots/county-pos-bar-2020-12-31.png" width="935" />

  - The top 20 counties according to estimated average percent change
    are shown. Counties at the top are experiencing the highest average
    growth rates in positive test results.
  - The average percent changes are estimated using county data in a
    log-linear model in order to capture exponential growth if an
    outbreak occurs.  
    <br /> <br /> <br />

<img src="plots/daily-re-line-2020-12-31.png" width="935" />

  - The instantaneous effective reproduction number, R<sub>e</sub> is an
    estimate of the average number of people that an infected person is
    spreading the virus to for that time window.  
  - Estimates *slighly lower* than 1 indicate that the spread of the
    virus is slowly decreasing. Estimates *slightly larger* than 1
    indicate a slow increasing spread of the virus. Estimates of *2 or
    greater* mean the virus is spreading rapidly.  
  - The shaded area shows the 80% credible interval where the true value
    is likely to be.
  - Further details on this metric can be found at a research
    [site](https://cbdrh.github.io/ozcoviz/#national-level-r_t) set-up
    by South Western Sydney Clinical School and the Centre for Big Data
    Research in Health.  
    <br /> <br /> <br />

<img src="plots/excess-death-col-2020-12-19.png" width="935" />

  - The
    [CDC’s](https://www.cdc.gov/nchs/nvss/vsrr/covid19/excess_deaths.htm)
    model uses data from 2015 to present to calculate estimates for
    weekly death counts. “Excess deaths” is defined as the difference
    between these estimates and the actual reported number of deaths.  
  - A portion of these excess deaths could be misclassified COVID-19
    deaths, “or potentially could be indirectly related to COVID-19
    (e.g., deaths from other causes occurring in the context of health
    care shortages or overburdened health care systems).”  
  - From the CDC site regarding the Causes-of-deaths data, “These causes
    were selected based on analyses of comorbid conditions reported on
    death certificates where COVID-19 was listed as a cause of death.
    Deaths with an underlying cause of death of COVID-19 are not
    included in these estimates of deaths due to other causes, but
    deaths where COVID-19 appeared on the death certificate as a
    multiple cause of death may be included in the cause-specific
    estimates.”


NEWS.md

## Changelog

  - 2020-04-12 - Georgia started separating non-state residents from
    their patient counts. Neither The New York Times nor Georgia
    adjusted the counts prior to the change. Without an adjustment, it
    destroyed the coherence of the data, so I’ve replaced them with
    South Carolina in the daily growth rate chart.

  - 2020-05-07 - The New York Times
    [decided](https://github.com/nytimes/covid-19-data/blob/master/PROBABLE-CASES-NOTE.md)
    to combine “probable” and “confirmed” positive cases and deaths in
    their datesets so that the data would remain consistent across all
    states. I think it provides a more accurate description of what’s
    going, so I’ve decided to keep their data source and discard Indiana
    State Department of Health’s (ISDH). The ramification is that many
    charts’ data will be a day late. Hopefully ISDH will opensource
    their probable cases and probable deaths, so the charts can return
    to being up-to-date.  

  - 2020-05-11 - The calculation of the rate for the Positive Test Rate
    chart was changed from using total counts to a rolling calculation
    over a 3 day windows. The interpretation has also changed based on
    the Johns Hopkins
    [article](https://coronavirus.jhu.edu/testing/testing-positivity).  

  - 2020-05-12 - Replaced the chart that compares daily growth rates and
    doubling times of states with similar population densities with a
    social distancing chart that uses Google Maps data. The daily growth
    rates chart ceased being interesting as doubling times and rates
    have pretty much plateaued. Also, I don’t want to present too many
    charts at once as it might create some information overload. If a
    second or third wave happens, then this chart might return.  

  - 2020-05-26 - Changed wording of the Hospitalizations - Ventilators -
    ICU Beds chart title. I misinterpreted the description of the
    hospitalizations data. I thought I was calculating the daily number
    of people being admitted to the hospital for COVID-19 when it was
    the change in the present count of people hospitalized for COVID-19.
    It’s still relevant because the governor’s speech used present count
    and not daily admittance as a guideline anyways. Apologees though.  

  - 2020-05-29 - County chart - Instead of using all the historic data
    to calculate average daily growth rates, I’m switching to only using
    data over the past 2 weeks which will make the estimates more
    sensitive to outbreaks if they happen.  

  - 2020-06-11 - Replaced the Apple Mobility chart with an OpenTables
    chart. Driving levels seem to have returned to pre-COVID levels, and
    with reopening stages well underway, it’ll be more useful to monitor
    how industries, like restaurants, hard-hit by COVID are recovering.
    The Apple Mobility chart will likely return if a second wave emerges
    though.  

  - 2020-06-13 - Changed the range and scale of the axes on the
    daily-positives-cumulative-policy chart. The data trend was no
    longer exponential so the log scaling isn’t really needed and it was
    bunching up the data points. At this point in the pandemic, what
    happens after the reopening stages is salient, so I’ve started the
    data at the point a couple weeks before stage 2.  

  - 2020-06-21 - Switched to the rt.live model for R<sub>e</sub>. I
    haven’t been confortable with estimate I’ve been using for awhile.
    Their estimate along with their uncertainty range look more
    reasonable than the one I was using before. They haven’t published
    their code yet, so I haven’t been able to dig into it too much. But
    the [faq](https://rt.live/faq) gives some hints to what they’re
    doing, and it seems sound.  

  - 2020-07-08 - Daily positive rate - During a
    [briefing](https://twitter.com/i/broadcasts/1ypKdwaqLegxW) I watched
    with Gov. Holcomb and Dr. Box, it was pretty clear that they pay
    close attention to positivity rates and use a seven day window to
    calculate theirs, so I’m switching my calculation window to 7 days.
    A 5% percent target was also emphasized by Dr. Box, so also changing
    my shaded region to a range from 0% - 5%.  

  - 2020-08-17 - Daily positive rate - I calculate the rate for all the
    data I have, but I don’t display the most recent two I’ve
    calculated, because there isn’t enough data to give a reasonable
    rate. I’ve decided remove another daily rate. So the last two rates
    that I calculate won’t be displayed; for the same reason. The lack
    of data makes these rates misleading, and some people don’t read the
    chart descriptions.  

  - 2020-09-18 - Daily positive rate, Median Age Weekly Cases and Tests,
    COVIDcast MSA Positivity - For about a month, ISDH’s dashboard has
    shown and Dr. Box has been referencing a much lower positivity rate,
    but haven’t included the testing data used to calculate that rate at
    Indiana Data Hub. Not sure when COVID\_TESTS\_ADMINISTRATED \[sic\]
    was added to the county-level dataset, but it’s there now and I’ll
    be using that field to calculate my positivity rate.
    
    The previous field I used for test counts only counted tests for
    unique individuals (i.e. no repeat tests of individuals were
    included). If a person was tested in March and then again in July,
    each test should be recorded for that month in order for the
    positivity rate to more accurately reflect the state of testing and
    viral spread. This new field includes these repeat tests and
    substantially increases the test counts which therefore lowers the
    positivity rate as a result.
    
    COVID\_TESTS\_ADMINISTRATED is an improvement, but it’s not ideal

R/hosp-icu-vent.R
# Monitors number of hospitalizations, availabilty of icu beds and ventilators


# Notes
# 1. The "triggers" are benchmarks for hospitalizations, icu beds, and ventilators that I'm monitoring. Discussed further in the readme.


# Sections
# 1. Set-up
# 2. hospitalizations: clean, calculate trigger
# 3. Hospitalizations chart
# 4. ICU, Vents: clean, calculate triggers
# 5. Gauge Plots
# 6. Trigger Plot
# 7. Combine all charts





##########
# Set-up
##########


pacman::p_load(extrafont, swatches, dplyr, tsibble, ggplot2, glue, patchwork, ggtext)

deep_rooted <- swatches::read_palette(glue("{rprojroot::find_rstudio_root_file()}/palettes/Deep Rooted.ase"))
purp_haz <- swatches::read_palette(glue("{rprojroot::find_rstudio_root_file()}/palettes/Purple Haze.ase"))
light_haz <- prismatic::clr_lighten(purp_haz, shift = 0.25)

# hospitalizations data
ct_dat_raw <- readr::read_csv("https://covidtracking.com/api/v1/states/daily.csv")
# beds, ventilators data
iv_dat_raw <- readr::read_csv(glue("{rprojroot::find_rstudio_root_file()}/data/beds-vents-complete.csv"))



##############################################
# hospitalizations: clean, calculate trigger
##############################################


# Get Indiana, cols = date, hospitalizedCurrently
ind_hosp <- ct_dat_raw %>% 
  filter(state == "IN" & hospitalizedCurrently != "NA") %>% 
  select(date, hospitalizedCurrently) %>% 
  mutate(date = lubridate::ymd(date)) %>%
  as_tsibble(index = date)

# current date of data
data_date <- ind_hosp %>% 
  as_tibble() %>%
  summarize(date = max(date)) %>% 
  pull(date)

# calc difference between one day and the previous day
hosp_changes <- ind_hosp %>% 
  as_tibble() %>% 
  mutate(hosp_diff = difference(hospitalizedCurrently)) %>% 
  arrange(desc(date)) %>% 
  pull(hosp_diff)

# calculate how many consecutive days of increasing/decreasing daily cases
count_consec_days <- function(x) {
  # rle: "run length encoding," counts runs of same value
  hosp_runs <- rle(sign(x))
  consec_days <- tibble(
    num_days = hosp_runs$lengths,
    sign = hosp_runs$values
  ) %>% 
    mutate(trend = case_when(sign == 1 ~ "increased",
                             sign == -1 ~ "decreased",
                             TRUE ~ "No change in")) %>% 
    slice(1) %>% 
    select(-sign)
}

consec_days <- count_consec_days(hosp_changes)



##########################
# Hospitalizations chart
##########################


# Trigger
# text styled depending on number of consecutive days and increasing or decreasing trend

# 1 day of decreased hospitalizations
neg_one <- glue("<b style='color: #33a532'>{consec_days$num_days[[1]]}</b> day of {consec_days$trend[[1]]} COVID-19 hospitializations")
# 1 day of increased hospitalizations
pos_one <- glue("<b style='color: #cf142b'>{consec_days$num_days[[1]]}</b> day of {consec_days$trend[[1]]} COVID-19 hospitializations")
# no change from yesterday
zero_days <- glue("{consec_days$trend[[1]]} COVID-19 hospitializations")
# between 2 and 6 days of increased hospitalizations
under_ft <- glue("<b style='color: #cf142b'>{consec_days$num_days[[1]]}</b> consecutive days of {consec_days$trend[[1]]} COVID-19 hospitializations")
# more than threshold of 14 days of increased hospitalizations
ft_over <- glue("<b style='color: #cf142b'>{consec_days$num_days[[1]]}</b> consecutive days of {consec_days$trend[[1]]} COVID-19 hospitializations <span style='font-family: \"Font Awesome 5 Free Solid\"; color: #cf142b'>&#xf06A;</span>")
# between 7 and 13 days of increased hospitalizations
ft_over_sev <- glue("<b style='color: #cf142b'>{consec_days$num_days[[1]]}</b> consecutive days of {consec_days$trend[[1]]} COVID-19 hospitializations <span style='font-family: \"Font Awesome 5 Free Solid\"; color: #ffae42'>&#xf071;</span>")
# more than 1 day of decreased hospitalizations
under_neg_one <- glue("<b style='color: #33a532'>{consec_days$num_days[[1]]}</b> consecutive days of {consec_days$trend[[1]]} COVID-19 hospitalizations")

# choose the subtitle text based number of consecutive days and trend light_haz[[2]]
trigger_dat_h <- consec_days %>% 
  mutate(text = case_when(num_days == 1 & trend == "increased" ~
                            pos_one,
                          num_days == 1 & trend == "decreased" ~
                            neg_one,
                          between(num_days, 2, 6) & trend == "increased" ~
                            under_ft,
                          between(num_days, 7, 13) & trend == "increased" ~
                            ft_over_sev,
                          num_days >= 14 & trend == "increased" ~
                            ft_over,
                          num_days > 1 & trend == "decreased" ~
                            under_neg_one,
                          TRUE ~ zero_days),
         type = "hosp")


caption_text <- glue("Last updated: {data_date}
                     Source: The COVID Tracking Project")

hosp_plot <- ggplot(data = ind_hosp,
                    aes(x = date, y = hospitalizedCurrently)) +
  geom_point(color = "#32a5a3") +
  geom_line(color = "#32a5a3") +
  expand_limits(y = c(min(ind_hosp$hospitalizedCurrently)-60, max(ind_hosp$hospitalizedCurrently) + 60),
                x = max(ind_hosp$date) + 8) +
  scale_x_date(date_breaks = "21 days",
               date_labels = "%b %d") +
  ggrepel::geom_label_repel(data = ind_hosp %>%
                              filter(date == max(date)),
                            aes(label = hospitalizedCurrently, size = 12),
                            nudge_x = 0, nudge_y = 18, point.padding = 1,
                            direction = "x", seed = 125) +
  geom_label(aes(x = as.Date("2020-04-24"),
                 y = max(hospitalizedCurrently) + 60,
                 label="Current COVID-19 Hospitalizations"),
             family="Roboto", fill = "black",
             size = 5, hjust = 0,
             label.size = 0, color = "white") +
  labs(x = NULL, y = NULL) +
  theme(text = element_text(family = "Roboto"),
        legend.position = "none",
        axis.text.x = element_text(color = "white",
                                   size = 11),
        axis.text.y = element_text(color = "white",
                                   size = 11),
        panel.background = element_rect(fill = "black",
                                        color = NA),
        plot.background = element_rect(fill = "black",
                                       color = NA),
        
        panel.border = element_blank(),
        panel.grid.minor = element_blank(),
        panel.grid.major = element_line(color = deep_rooted[[7]]))




#########################################
# ICU, Vents: clean, calculate triggers
#########################################


# calcs percentages, assigns 1 for above threshold, -1 for below (needed for sign() fun below)
iv_dat <- iv_dat_raw %>%
  mutate(icu_beds_pct = beds_available_icu_beds_total/beds_icu_total,
         vent_pct = vents_all_available_vents_not_in_use/vents_total,
         icu_status = ifelse(icu_beds_pct >= 0.400, 1, -1),
         vent_status = ifelse(vent_pct >= 0.700, 1, -1)) %>% 
  arrange(desc(date))

# count consecutive days above and below threshold
count_consec_days_iv <- function(x) {
  # rle: "run length encoding," counts runs of same value
  iv_runs <- rle(sign(x))
  consec_days <- tibble(
    num_days = iv_runs$lengths,
    sign = iv_runs$values
  ) %>% 
    mutate(trend = case_when(sign == 1 ~ "above",
                             sign == -1 ~ "below",
                             TRUE ~ "no change in")) %>% 
    slice(1) %>% 
    select(-sign)
}

consec_days_i <- count_consec_days_iv(iv_dat$icu_status)
consec_days_v <- count_consec_days_iv(iv_dat$vent_status)



# text styled depending on number of consecutive days and above or below trend

# ICU Beds text
# 1 day above 40%
pos_one_i <- glue("<b style='color: #33a532'>{consec_days_i$num_days[[1]]}</b> day {consec_days_i$trend[[1]]} 40% availability for ICU beds")
# 1 day below 40%
neg_one_i <- glue("<b style='color: #cf142b'>{consec_days_i$num_days[[1]]}</b> day {consec_days_i$trend[[1]]} 40% availability for ICU beds")
# 0 days == no change
zero_days_i <- glue("{consec_days_i$trend[[1]]} in availability of ICU beds from yesterday")
# between 2 and 6 days below 40%
under_ft_i <- glue("<b style='color: #cf142b'>{consec_days_i$num_days[[1]]}</b> consecutive days of being {consec_days_i$trend[[1]]} 40% availability for ICU beds")
# 14 days and over days below 40%
ft_over_i <- glue("<b style='color: #cf142b'>{consec_days_i$num_days[[1]]}</b> consecutive days of being {consec_days_i$trend[[1]]} 40% availability for ICU beds <span style='font-family: \"Font Awesome 5 Free Solid\"; color: #cf142b'>&#xf06A;</span>")
# between 7 and 13 days below 40%
ft_over_i_sev <- glue("<b style='color: #cf142b'>{consec_days_i$num_days[[1]]}</b> consecutive days of being {consec_days_i$trend[[1]]} 40% availability for ICU beds <span style='font-family: \"Font Awesome 5 Free Solid\"; color: #ffae42'>&#xf071;</span>")
# more than 1 day above 40%
above_pos_one_i <- glue("<b style='color: #33a532'>{consec_days_i$num_days[[1]]}</b> consecutive days of being {consec_days_i$trend[[1]]} 40% availability for ICU beds")

# choose the subtitle text based number of consecutive days and trend
trigger_dat_i <- consec_days_i %>% 
  mutate(text = case_when(num_days == 1 & trend == "above" ~
                            pos_one_i,
                          num_days == 1 & trend == "below" ~
                            neg_one_i,
                          between(num_days, 2, 6) & trend == "below" ~
                            under_ft_i,
                          between(num_days, 7, 13) & trend == "below" ~
                            ft_over_i_sev,
                          num_days >= 14 & trend == "below" ~
                            ft_over_i,
                          num_days > 1 & trend == "above" ~
                            above_pos_one_i,
                          TRUE ~ zero_days_i),
         type = "icu")



# Ventilators text
# 1 day above 70%
pos_one_v <- glue("<b style='color: #33a532'>{consec_days_v$num_days[[1]]}</b> day {consec_days_v$trend[[1]]} 70% availability for ventilators")
# 1 day below 70%
neg_one_v <- glue("<b style='color: #cf142b'>{consec_days_v$num_days[[1]]}</b> day {consec_days_v$trend[[1]]} 70% availability for ventilators")
# 0 days == no change
zero_days_v <- glue("{consec_days_v$trend[[1]]} in availability of ventilators from yesterday")
# between 2 and 6 days below 70%
under_ft_v <- glue("<b style='color: #cf142b'>{consec_days_v$num_days[[1]]}</b> consecutive days of being {consec_days_v$trend[[1]]} 70% availability for ventilators")
# 14 and over days below 70%
ft_over_v <- glue("<b style='color: #cf142b'>{consec_days_v$num_days[[1]]}</b> consecutive days of being {consec_days_v$trend[[1]]} 70% availability for ventilators <span style='font-family: \"Font Awesome 5 Free Solid\"; color: #cf142b'>&#xf06A;</span>")
# betweeen 7 and 13 days below 70%
ft_over_v_sev <- glue("<b style='color: #cf142b'>{consec_days_v$num_days[[1]]}</b> consecutive days of being {consec_days_v$trend[[1]]} 70% availability for ventilators <span style='font-family: \"Font Awesome 5 Free Solid\"; color: #ffae42'>&#xf071;</span>")
# more than 1 day above 70%
above_pos_one_v <- glue("<b style='color: #33a532'>{consec_days_v$num_days[[1]]}</b> consecutive days of being {consec_days_v$trend[[1]]} 70% availability for ventilators")

# choose the subtitle text based number of consecutive days and trend
trigger_dat_v<- consec_days_v %>% 
  mutate(text = case_when(num_days == 1 & trend == "above" ~
                            pos_one_v,
                          num_days == 1 & trend == "below" ~
                            neg_one_v,
                          between(num_days, 2, 6) & trend == "below" ~
                            under_ft_v,
                          between(num_days, 7, 13) & trend == "below" ~
                            ft_over_v_sev,
                          num_days >= 14 & trend == "below" ~
                            ft_over_v,
                          num_days > 1 & trend == "above" ~
                            above_pos_one_v,
                          TRUE ~ zero_days_v),
         type = "vent")



################
# Gauge Plots
################

# used for both gauge plots
# cols: type (icu or vent), percent, title, label (text of percent col)
gauge_dat <- iv_dat %>% 
  slice(1) %>% 
  select(icu_beds_pct, vent_pct) %>% 
  tidyr::pivot_longer(cols = everything(),
                      names_to = "type",
                      values_to = "percent") %>% 
  mutate(title = c("ICU Beds Available", "Ventilators Available"),
         label = scales::percent(percent, accuracy = 0.1))


icu_gauge <- ggplot(data = gauge_dat %>% 
                      filter(type == "icu_beds_pct")) +
  # green bar
  geom_rect(aes(ymax=1, ymin=0, xmax=2, xmin=1), fill ="#33a532") +
  # red bar that overlays the green bar
  geom_rect(aes(fill = type, ymax = 0.4, ymin = 0, xmax = 2, xmin = 1)) + 
  # bends bars to a half circle
  coord_polar(theta = "y",start=-pi/2) + xlim(c(0, 2)) + ylim(c(0,2)) +
  # segments the bar between red and green; black blends with black background
  geom_hline(aes(yintercept = 0.40),
             color = "black", size = 1.3) +
  # the "needle"
  geom_hline(aes(yintercept = percent),
             color = "white", size = 1.3) +
  # percent text
  geom_text(aes(x = 0.85, y = 1.5, label = label), colour="white", size=6.5, fontface = "bold", family = "Roboto") +
  # title depending on type
  geom_text(aes(x=1.5, y=1.5, label=title), size=6.2, family = "Roboto", color = "white") +
  theme_void() +
  # not sure why there's two values
  scale_fill_manual(values = c("#cf142b", "#cf142b")) +
  # maybe this the stroke for the bars?
  scale_colour_manual(values = c("#cf142b", "#cf142b")) +
  theme(legend.position = "none",
        axis.ticks = element_blank(),
        axis.line = element_blank(),
        axis.text = element_blank(),
        panel.background = element_rect(fill = "black",
                                        color = NA),
        plot.background = element_rect(fill = "black",
                                       color = NA))



vents_gauge <- ggplot(data = gauge_dat %>% 
                        filter(type == "vent_pct")) +
  geom_rect(aes(ymax=1, ymin=0, xmax=2, xmin=1), fill ="#33a532") +
  geom_rect(aes(fill = type, ymax = 0.7, ymin = 0, xmax = 2, xmin = 1)) + 
  coord_polar(theta = "y",start=-pi/2) + xlim(c(0, 2)) + ylim(c(0,2)) +
  geom_hline(aes(yintercept = 0.70),
             color = "black", size = 1.3) +
  geom_hline(aes(yintercept = percent),
             color = "white", size = 1.3) +
  geom_text(aes(x = 0.85, y = 1.5, label = label), colour="white", size=6.5, fontface = "bold", family = "Roboto") +
  geom_text(aes(x=1.5, y=1.5, label=title), size=6.2, family = "Roboto", color = "white") +
  theme_void() +
  scale_fill_manual(values = c("#cf142b", "#cf142b")) +
  scale_colour_manual(values = c("#cf142b", "#cf142b")) +
  theme(legend.position = "none",
        axis.ticks = element_blank(),
        axis.line = element_blank(),
        axis.text = element_blank(),
        panel.background = element_rect(fill = "black",
                                        color = NA),
        plot.background = element_rect(fill = "black",
                                       color = NA))



################
# Trigger plot
################



status_dat <- trigger_dat_h %>%
  bind_rows(trigger_dat_i) %>%
  bind_rows(trigger_dat_v)

status_plot <- ggplot(status_dat, aes(y = text)) +
  ggtext::geom_richtext(data = status_dat %>% 
                          slice(1), 
                        aes(label= text,
                           x = 0,y = 0.8,
                           label.color = NA),
                        fill = "black",
                        color = "white",
                        size = 5, hjust = "left") +
  ggtext::geom_richtext(data = status_dat %>% 
                          slice(2), 
                        aes(label= text,
                           x = 0, y = 0.50,
                           label.color = NA),
                        fill = "black",
                        color = "white",
                        size = 5, hjust = "left") +
  ggtext::geom_richtext(data = status_dat %>% 
                          slice(3), 
                        aes(label= text,
                           x = 0, y = 0.20,
                           label.color = NA),
                        fill = "black",
                        color = "white",
                        size = 5, hjust = "left") +
  # sets the range of the grid, so you have some idea of coord system for text.
xlim(0, 1) + ylim(0, 1) +
  theme_void() +
  theme(panel.background = element_rect(fill = "black",
                                        color = NA),
        panel.border = element_blank(),
        plot.background = element_rect(fill = "black",
                                       color = NA))



######################
# Combine all charts
######################

# Pay attention to groupings by parentheses
# 1st plot_layout affects left col; 2nd plot_layout affects entire chart
# see gtable notes for "null" explanation
all_charts <- ((hosp_plot/status_plot + plot_layout(heights = unit(c(13, 1), c('cm', 'null')))) | 
                 (icu_gauge/vents_gauge)) +
  plot_layout(widths = c(2,1)) +
  plot_annotation(title = "Tracking COVID-19 Hospitalizations, ICU and Ventilator Availability",
                  subtitle = glue("Last updated: {data_date}"),
                  caption = glue("Sources: The Indiana Data Hub
                                 The COVID Tracking Project")) &
  theme(plot.title = element_text(color = "white",
                                  size = 16,
                                  family = "Roboto"),
        plot.subtitle = element_text(color = "white",
                                     size = 14,
                                     family = "Roboto"),
        plot.caption = element_text(color = "white",
                                    size = 12,
                                    family = "Roboto"),
        panel.background = element_rect(fill = "black",
                                        color = NA),
        panel.border = element_blank(),
        plot.background = element_rect(fill = "black",
                                       color = NA))


plot_path <- glue("{rprojroot::find_rstudio_root_file()}/plots/hosp-icu-vent-{data_date}.png")

ggsave(plot_path, plot = all_charts, dpi = "screen", width = 33, height = 20, units = "cm")

